llap-tez/src/java/org/apache/hadoop/hive/llap/tezplugins/metrics/LlapMetricsCollector.java

```diff
@@ -0,0 +1,194 @@
+import org.apache.hadoop.hive.llap.daemon.rpc.LlapDaemonProtocolProtos;
+import org.apache.hadoop.hive.llap.impl.LlapManagementProtocolClientImpl;
+import org.apache.hadoop.hive.llap.registry.LlapServiceInstance;
+import org.apache.hadoop.hive.registry.ServiceInstanceStateChangeListener;
+import org.apache.hadoop.io.retry.RetryPolicies;
+import org.apache.hadoop.io.retry.RetryPolicy;
+import org.apache.hadoop.net.NetUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.net.SocketFactory;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Collect metrics from the llap daemons in every given milliseconds.
+ */
+public class LlapMetricsCollector implements ServiceInstanceStateChangeListener<LlapServiceInstance> {
+
+  private static final Logger LOG = LoggerFactory.getLogger(LlapMetricsCollector.class);
+  private static final String THREAD_NAME = "LlapTaskSchedulerMetricsCollectorThread";
+  private static final long INITIAL_DELAY_MSEC = 60000L;
+
+  private final Configuration conf;
+  private final ScheduledExecutorService scheduledMetricsExecutor;
+  private final LlapManagementProtocolClientImplFactory clientFactory;
+  private final Map<String, LlapManagementProtocolClientImpl> llapClients;
+  private final Map<String, LlapMetrics> instanceStatisticsMap;
+  private final long metricsCollectionMs;
+
+
+  public LlapMetricsCollector(Configuration conf) {
+    this(
+            conf,
+            Executors.newSingleThreadScheduledExecutor(
+                    new ThreadFactoryBuilder().setDaemon(true).setNameFormat(THREAD_NAME)
+                            .build()),
+            LlapManagementProtocolClientImplFactory.basicInstance(conf));
+  }
+
+  @VisibleForTesting
+  LlapMetricsCollector(Configuration conf, ScheduledExecutorService scheduledMetricsExecutor,
+                       LlapManagementProtocolClientImplFactory clientFactory) {
+    this.conf = conf;
+    this.scheduledMetricsExecutor = scheduledMetricsExecutor;
+    this.clientFactory = clientFactory;
+    this.llapClients = new HashMap<>();
+    this.instanceStatisticsMap = new ConcurrentHashMap<>();
+    this.metricsCollectionMs = HiveConf.getTimeVar(conf,
+            HiveConf.ConfVars.LLAP_TASK_SCHEDULER_AM_COLLECT_DAEMON_METRICS_MS, TimeUnit.MILLISECONDS);
+  }
+
+  public void start() {
+    if (enabled()) {
+      scheduledMetricsExecutor.scheduleAtFixedRate(() -> {
+        collectMetrics();
+      }, INITIAL_DELAY_MSEC, metricsCollectionMs, TimeUnit.MILLISECONDS);
+    }
+  }
+
+  private boolean enabled() {
+    return metricsCollectionMs > 0;
+  }
+
+  public void shutdown() {
+    scheduledMetricsExecutor.shutdownNow();
+  }
+
+  @VisibleForTesting
+  void collectMetrics() {
+    if (enabled()) {
```

Comment:
Unnecessary check: The thread, calling collectMetrics is never created if ms <= 0.

Reply:
Fixed.